server/sonar-server/src/main/java/org/sonar/server/computation/ws/QualityGateStatusAction.java

```diff
@@ -0,0 +1,167 @@
+    WebService.NewAction action = controller.createAction("quality_gate_status")
+      .setDescription("Quality gate status for a given CE task. <br />" +
+        "Returns a http code 404 if the snapshot associated with the task is not found.<br />" +
+        "Requires 'Administer System' or 'Execute Analysis' permission.")
+      .setResponseExample(getClass().getResource("quality_gate_status-example.json"))
+      .setSince("5.3")
+      .setInternal(true)
+      .setHandler(this);
+
+    action.createParam("taskId")
+      .setDescription("Task id")
+      .setRequired(true)
+      .setExampleValue(Uuids.UUID_EXAMPLE_03);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    QualityGateStatusWsRequest qualityGateStatusWsRequest = toQualityGateStatusWsRequest(request);
+    QualityGateStatusWsResponse qualityGateStatusWsResponse = doHandle(qualityGateStatusWsRequest);
+
+    writeProtobuf(qualityGateStatusWsResponse, request, response);
+  }
+
+  private QualityGateStatusWsResponse doHandle(QualityGateStatusWsRequest request) {
+    checkScanOrAdminPermission();
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      String taskUuid = request.getTaskId();
+      CeActivityDto ceActivityDto = getCeActivity(dbSession, taskUuid);
+      SnapshotDto snapshotDto = getSnapshot(dbSession, ceActivityDto.getSnapshotId(), taskUuid);
+      MeasureDto measureDto = getQualityGateDetailsMeasure(dbSession, snapshotDto, taskUuid);
+
+      return buildResponse(measureDto.getData(), taskUuid);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private CeActivityDto getCeActivity(DbSession dbSession, String taskUuid) {
+    return checkFoundWithOptional(dbClient.ceActivityDao().selectByUuid(dbSession, taskUuid),
+      "Task with id '%s' is not found", taskUuid);
+  }
+
+  private MeasureDto getQualityGateDetailsMeasure(DbSession dbSession, SnapshotDto snapshotDto, String taskUuid) {
+    List<MeasureDto> measures = dbClient.measureDao().selectBySnapshotIdAndMetricKeys(snapshotDto.getId(),
+      Collections.singleton(CoreMetrics.QUALITY_GATE_DETAILS_KEY), dbSession);
+    if (measures.isEmpty()) {
+      throw new NotFoundException(format("Quality gate details of task with id '%s' are not found", taskUuid));
+    }
+
+    return measures.get(0);
+  }
+
+  private SnapshotDto getSnapshot(DbSession dbSession, @Nullable Long snapshotId, String taskId) {
+    SnapshotDto snapshotDto = (snapshotId == null)
+      ? null
+      : dbClient.snapshotDao().selectById(dbSession, snapshotId);
+
+    return checkFound(snapshotDto, "Snapshot for task id '%s' is not found", taskId);
+  }
+
+  private static QualityGateStatusWsResponse buildResponse(String measureData, String taskUuid) {
+    try {
+      QualityGateDetails.Builder responseBuilder = QualityGateDetails.newBuilder();
+      JsonParser parser = new JsonParser();
```

Comment:
whatever the usage, the code that reads and writes a structure in database must be in the same place.